onnxruntime/core/framework/data_types.cc

```diff
@@ -369,6 +381,52 @@ MLDataType TensorTypeBase::Type() {
   return &tensor_base;
 }
 
+/// SparseTensor
+
+struct SparseTensorTypeBase::Impl : public data_types_internal::TypeProtoImpl {
+};
+
+SparseTensorTypeBase::SparseTensorTypeBase() : impl_(new Impl()) {}
+SparseTensorTypeBase::~SparseTensorTypeBase() {
+  delete impl_;
+}
+
+bool SparseTensorTypeBase::IsCompatible(const ONNX_NAMESPACE::TypeProto& type_proto) const {
+  const auto* thisProto = GetTypeProto();
+  if (&type_proto == thisProto) {
+    return true;
+  }
+  if (type_proto.value_case() != TypeProto::ValueCase::kSparseTensorType) {
+    return false;
+  }
+
+  ORT_ENFORCE(thisProto->value_case() == TypeProto::ValueCase::kSparseTensorType);
+  ORT_ENFORCE(thisProto->sparse_tensor_type().has_elem_type());
+
```

Comment:
Do we need to check these here? I would have expected we validated these conditions once when constructing the class instead.

Reply:
I think you are right. I am following the same style as in the preceding TensorTypeBase::IsCompatible (and all other ::IsCompatible as well). I assume that these are sanity-checks (which may often be redundant). I will leave them in for uniformity for now.